pkg/sign/transparency.go

```diff
@@ -0,0 +1,140 @@
+	"github.com/sigstore/rekor/pkg/client"
+	"github.com/sigstore/rekor/pkg/generated/client/entries"
+	"github.com/sigstore/rekor/pkg/generated/models"
+	"github.com/sigstore/rekor/pkg/pki"
+	"github.com/sigstore/rekor/pkg/tle"
+	"github.com/sigstore/rekor/pkg/types"
+	"github.com/sigstore/rekor/pkg/types/dsse"
+	"github.com/sigstore/rekor/pkg/types/hashedrekord"
+	"github.com/sigstore/rekor/pkg/util"
+
+	// To initialize rekor types
+	_ "github.com/sigstore/rekor/pkg/types/dsse/v0.0.1"
+	_ "github.com/sigstore/rekor/pkg/types/hashedrekord/v0.0.1"
+)
+
+type Transparency interface {
+	GetTransparencyLog([]byte, *protobundle.Bundle) error
+}
+
+type Rekor struct {
+	options *RekorOptions
+}
+
+type RekorOptions struct {
+	// URL of Fulcio instance
+	BaseURL string
+	// Optional timeout for network requests
+	Timeout time.Duration
+	// Optional version string for user agent
+	LibraryVersion string
+}
+
+func NewRekor(opts *RekorOptions) *Rekor {
+	return &Rekor{options: opts}
+}
+
+func (r *Rekor) GetTransparencyLog(pubKeyPEM []byte, b *protobundle.Bundle) error {
+	artifactProperties := types.ArtifactProperties{
+		PublicKeyBytes: [][]byte{pubKeyPEM},
+	}
+
+	dsseEnvelope := b.GetDsseEnvelope()
+	messageSignature := b.GetMessageSignature()
+	verificationMaterial := b.GetVerificationMaterial()
+	bundleCertificate := verificationMaterial.GetCertificate()
+
+	var proposedEntry models.ProposedEntry
+
+	switch {
+	case dsseEnvelope != nil:
+		dsseType := dsse.New()
+
+		artifactBytes, err := json.Marshal(dsseEnvelope)
+		if err != nil {
+			return err
+		}
+
+		artifactProperties.ArtifactBytes = artifactBytes
+
+		proposedEntry, err = dsseType.CreateProposedEntry(context.TODO(), "", artifactProperties)
+		if err != nil {
+			return err
+		}
+	case messageSignature != nil:
+		hashedrekordType := hashedrekord.New()
+
+		if bundleCertificate == nil {
+			return errors.New("hashedrekord requires X.509 certificate")
+		}
+
+		hexDigest := hex.EncodeToString(messageSignature.MessageDigest.Digest)
+
+		artifactProperties.PKIFormat = string(pki.X509)
+		artifactProperties.SignatureBytes = messageSignature.Signature
+		artifactProperties.ArtifactHash = util.PrefixSHA(hexDigest)
+
+		var err error
+		proposedEntry, err = hashedrekordType.CreateProposedEntry(context.TODO(), "", artifactProperties)
+		if err != nil {
+			return err
+		}
+	default:
+		return errors.New("unable to find signature in bundle")
+	}
+
+	params := entries.NewCreateLogEntryParams()
+	if r.options.Timeout > 0 {
+		params.SetTimeout(r.options.Timeout)
+	}
+	params.SetProposedEntry(proposedEntry)
+
+	client, err := client.GetRekorClient(r.options.BaseURL, client.WithUserAgent(constructUserAgent(r.options.LibraryVersion)))
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Entries.CreateLogEntry(params)
+	if err != nil {
+		return err
+	}
+
+	entry := resp.Payload[resp.ETag]
+	tlogEntry, err := tle.GenerateTransparencyLogEntry(entry)
```

Comment:
Unrelated to this PR, I think I noticed a bug - https://github.com/sigstore/rekor/blob/92cf4d5d3682ffcbf4138341f1bc5f7a66766549/pkg/tle/tle.go#L88 doesn't base64 decode the SET before setting it, which is different than what you had here. Can you verify if this is a bug?

Reply:
Hm, maybe not? https://github.com/sigstore/cosign/blob/121115774e8c662648165f2924e7354292365ae8/pkg/cosign/tlog.go#L492 doesn't base64-decode it.

Reply:
I did generate a bundle with these changes, and it verified with a policy that required an observer timestamp (and it didn't have a timestamp authority entry), so I believe this is working:

```
$ cat bundle.json | jq -r ".verificationMaterial.tlogEntries[0].inclusionPromise.signedEntryTimestamp" | base64 -d
0E J??s??U?P??7????43?$D
                        <s;?b???!ԟ?:
                                    ?Ljj-K??>S?j??mv??????*	%
```

Reply:
If I were to guess, during marshalling, it doesn't re-encode it as base64. I'll play around with it offline, but if this is working, perfect.

Reply:
Just to close out on this, the difference between your previous change and this was that you base64 decoded `e.Verification.SignedEntryTimestamp.String()`, which outputs a base64-encoded SET. Without calling `MarshalJSON` or `String`, you're working on the bytes.